File: 2022/day6/day6.go
package day6

import (
	"fmt"
	"log"
	"strings"

	util "github.com/blaine-t-bush/advent-of-code-2022/util"
)

func sliceContainsCharacter(chars []string, check string) bool {
	for _, char := range chars {
		if char == check {
			return true
		}
	}

	return false
}

func charactersAreUnique(chars []string) bool {
	for i, char := range chars {
		filteredChars := []string{}
		filteredChars = append(filteredChars, chars[:i]...)
		if i < len(chars)-1 {
			filteredChars = append(filteredChars, chars[i+1:]...)
		}
		if sliceContainsCharacter(filteredChars, char) {
			return false
		}
	}

	return true
}

// get index of first character after start-of-packet marker,
// which is any sequence of four unique characters.
func getFirstMarker(text string, uniqueChars int) int {
	length := len(text)
	for i := 0; i <= length-uniqueChars+1; i++ {
		if charactersAreUnique(strings.Split(text[i:i+uniqueChars], "")) {
			return i + uniqueChars - 1
		}
	}

	log.Fatal("could not find start marker")
	return -1
}

func SolvePartOne() int {
	signal := util.ReadInput("./2022/day6/input.txt")[0]
	firstStartMarkerIndex := getFirstMarker(signal, 4) + 1
	fmt.Println(firstStartMarkerIndex)
	return firstStartMarkerIndex
}

func SolvePartTwo() int {
	signal := util.ReadInput("./2022/day6/input.txt")[0]
	firstStartMarkerIndex := getFirstMarker(signal, 14) + 1
	fmt.Println(firstStartMarkerIndex)
	return firstStartMarkerIndex
}

File: 2016/day6.py
def get_most_common_character(strings, index: int):
    occurrences = {}
    for string in strings:
        char = string[index]
        if char in occurrences:
            occurrences[char] = occurrences[char] + 1
        else:
            occurrences[char] = 1

    return max(occurrences, key=occurrences.get)


def get_least_common_character(strings, index: int):
    occurrences = {}
    for string in strings:
        char = string[index]
        if char in occurrences:
            occurrences[char] = occurrences[char] + 1
        else:
            occurrences[char] = 1

    return min(occurrences, key=occurrences.get)


def solve_part_one():
    with open("./inputs/day6.txt") as f:
        strings = [s.rstrip("\n") for s in f.readlines()]

    code = ""
    for i in range(len(strings[0])):
        code += get_most_common_character(strings, i)

    return code


def solve_part_two():
    with open("./inputs/day6.txt") as f:
        strings = [s.rstrip("\n") for s in f.readlines()]

    code = ""
    for i in range(len(strings[0])):
        code += get_least_common_character(strings, i)

    return code


print(f"Part 1: message '{solve_part_one()}'")
print(f"Part 2: message '{solve_part_two()}'")
